File: Examples/VisualBasic/3DScene/Save3DMeshesInCustomBinaryFormat.vb
Imports Aspose.ThreeD
Imports System.IO
Imports Aspose.ThreeD.Entities
Imports Aspose.ThreeD.Utilities

Public Class Save3DMeshesInCustomBinaryFormat
    ' ExStart:Save3DMeshesInCustomBinaryFormat
    Public Shared Sub Run()
        ' The path to the documents directory.
        Dim MyDir As String = RunExamples.GetDataDir()
        ' load a 3D file
        Dim scene As New Scene(MyDir + "test.fbx")
        ' visit each descent nodes
        scene.RootNode.Accept(AddressOf VisitMeshNodes)
    End Sub
    Private Shared Function VisitMeshNodes(node As Node)
        ' The path to the documents directory.
        Dim MyDir As String = RunExamples.GetDataDir()
        ' open file for writing in binary mode
        Using writer = New BinaryWriter(New FileStream(MyDir + "Save3DMeshesInCustomBinaryFormat_out", FileMode.Create, FileAccess.Write))
            For Each entity As Entity In node.Entities
                ' only convert meshes, lights/camera and other stuff will be ignored
                If Not (TypeOf entity Is IMeshConvertible) Then
                    Continue For
                End If
                Dim m As Mesh = DirectCast(entity, IMeshConvertible).ToMesh()
                Dim controlPoints = m.ControlPoints
                ' triangulate the mesh, so triFaces will only store triangle indices
                Dim triFaces As Integer()() = PolygonModifier.Triangulate(controlPoints, m.Polygons)
                ' gets the global transform matrix
                Dim transform As Matrix4 = node.GlobalTransform.TransformMatrix
                ' write number of control points and triangle indices
                writer.Write(controlPoints.Count)
                writer.Write(triFaces.Length)
                ' write control points
                For i As Integer = 0 To controlPoints.Count - 1
                    ' calculate the control points in world space and save them to file
                    Dim cp = transform * controlPoints(i)
                    writer.Write(CSng(cp.x))
                    writer.Write(CSng(cp.y))
                    writer.Write(CSng(cp.z))
                Next
                ' write triangle indices
                For i As Integer = 0 To triFaces.Length - 1
                    writer.Write(triFaces(i)(0))
                    writer.Write(triFaces(i)(1))
                    writer.Write(triFaces(i)(2))
                Next
            Next
            Return True
        End Using
    End Function
    ' ExEnd:Save3DMeshesInCustomBinaryFormat
End Class

File: Examples/CSharp/3DScene/Save3DMeshesInCustomBinaryFormat.cs
using Aspose.ThreeD;
using Aspose.ThreeD.Entities;
using Aspose.ThreeD.Utilities;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Aspose._3D.Examples.CSharp._3DScene
{
    class Save3DMeshesInCustomBinaryFormat
    {
        public static void Run()
        {
            // ExStart:Save3DMeshesInCustomBinaryFormat
            // load a 3D file
            Scene scene = new Scene(RunExamples.GetDataFilePath("test.fbx"));

            /*
            * 3D format demonstration is simple
            * 
            * struct File {
            *   MeshBlock blocks[];
            * };
            *
            * struct Vertex {
            *   float x;
            *   float y;
            *   float z;
            * };
            * 
            * struct Triangle {
            *   int a;
            *   int b;
            *   int c;
            * };
            * 
            * struct MeshBlock {
            *   int numControlPoints;
            *   int numTriangles;
            *   Vertex vertices[numControlPoints];
            *   Triangle faces[numTriangles];
            * };
            */

            // open file for writing in binary mode
            using (var writer = new BinaryWriter(new FileStream(RunExamples.GetOutputFilePath("Save3DMeshesInCustomBinaryFormat_out"), FileMode.Create, FileAccess.Write)))
            {
                // visit each descent nodes
                scene.RootNode.Accept(delegate(Node node)
                {
                    foreach (Entity entity in node.Entities)
                    {
                        // only convert meshes, lights/camera and other stuff will be ignored
                        if (!(entity is IMeshConvertible))
                            continue;
                        Mesh m = ((IMeshConvertible)entity).ToMesh();
                        var controlPoints = m.ControlPoints;
                        // triangulate the mesh, so triFaces will only store triangle indices
                        int[][] triFaces = PolygonModifier.Triangulate(controlPoints, m.Polygons);
                        // gets the global transform matrix
                        Matrix4 transform = node.GlobalTransform.TransformMatrix;

                        // write number of control points and triangle indices
                        writer.Write(controlPoints.Count);
                        writer.Write(triFaces.Length);
                        // write control points
                        for (int i = 0; i < controlPoints.Count; i++)
                        {
                            // calculate the control points in world space and save them to file
                            var cp = transform * controlPoints[i];
                            writer.Write((float)cp.x);
                            writer.Write((float)cp.y);
                            writer.Write((float)cp.z);
                        }
                        // write triangle indices
                        for (int i = 0; i < triFaces.Length; i++)
                        {
                            writer.Write(triFaces[i][0]);
                            writer.Write(triFaces[i][1]);
                            writer.Write(triFaces[i][2]);
                        }
                    }
                    return true;
                });
            }
            // ExEnd:Save3DMeshesInCustomBinaryFormat
        }
    }
}
